Increase asyncify stack size

microbit-matt-hillsdon opened the pull request:
This fixes the example on #105.

I'll do some testing to find a reasonable value in a bit.

dpgeorge commented:
I use the following test to determine maximum stack depth when calling a function:
```
def g():
    global depth
    depth += 1
    g()
depth = 0
try:
    g()
except RuntimeError:
    pass
print('maximum recursion depth g():', depth)
```
That works only when `MICROPY_STACK_CHECK` is enabled, which it's not for this simulator.  It might be worth trying to enable that option here, which would protect against large call depths.

Otherwise, you can make a simple fixed-depth test like this:
```
def f(n):
    if n:
        f(n - 1)
f(168)
```

microbit-matt-hillsdon commented:
> I use the following test to determine maximum stack depth when calling a function:
> 
> ```
> def g():
>     global depth
>     depth += 1
>     g()
> depth = 0
> try:
>     g()
> except RuntimeError:
>     pass
> print('maximum recursion depth g():', depth)
> ```
> 
> That works only when `MICROPY_STACK_CHECK` is enabled, which it's not for this simulator. It might be worth trying to enable that option here, which would protect against large call depths.
> 
> Otherwise, you can make a simple fixed-depth test like this:
> 
> ```
> def f(n):
>     if n:
>         f(n - 1)
> f(168)
> ```

Yeah that makes sense. I got fixated on the structure of the original example. I've updated it to be a Python example alongside the others. Both examples you give work for me without enabling MICROPY_STACK_CHECK.

If I do enable MICROPY_STACK_CHECK (which is enabled for the CODAL port) I see worse behaviour:

```
FATAL: uncaught NLR 0x510d30 [firmware.js:1483:35](http://localhost:63072/build/firmware.js)
Aborted(Assertion failed: native function `stackRestore` called after runtime exit (use NO_EXIT_RUNTIME to keep it alive after main() exits)) [firmware.js:990:6](http://localhost:63072/build/firmware.js)
Aborted(Assertion failed: native function `mp_js_force_stop` called after runtime exit (use NO_EXIT_RUNTIME to keep it alive after main() exits)) [firmware.js:990:6](http://localhost:63072/build/firmware.js)
Simulator internal error: [demo:205:23](http://localhost:63072/demo)
Error: Aborted(Assertion failed: native function `mp_js_force_stop` called after runtime exit (use NO_EXIT_RUNTIME to keep it alive after main() exits))
    abort http://localhost:63072/build/firmware.js:1013
    assert http://localhost:63072/build/firmware.js:512
    createModule http://localhost:63072/build/firmware.js:1075
    forceStop http://localhost:63072/build/simulator.js:2201
    createRunningPromise http://localhost:63072/build/simulator.js:2470
    start http://localhost:63072/build/simulator.js:2440
    flash http://localhost:63072/build/simulator.js:2548
    createMessageListener http://localhost:63072/build/simulator.js:2751
    EventListener.handleEvent* http://localhost:63072/build/simulator.js:2826
    <anonymous> http://localhost:63072/build/simulator.js:2827
[demo:206:23](http://localhost:63072/demo)
```

Some of that (the mp_js_force_stop bit at least) seems like it might be a knock-on error from whatever originally goes wrong.

For the moment at least, I think it makes sense to go with the PR as I've just updated it.

dpgeorge commented on src/demo.html:
```diff
@@ -100,6 +100,7 @@ <h1>MicroPython-micro:bit simulator example embedding</h1>
               </option>
               <option value="sound_effects_user">Sound effects (user)</option>
               <option value="speech">Speech</option>
+              <option value="stack_size">Stack size</option>
```
What does this do?

microbit-matt-hillsdon replied:
Adds the example to the list in [the demo](https://review-python-simulator.usermbit.org/asyncify-stack-size/demo.html)

![image](https://github.com/microbit-foundation/micropython-microbit-v2-simulator/assets/44397098/ff15d21d-9c4f-4d31-bd82-669604389949)

dpgeorge replied:
Ah, very nice!
